### PR TITLE
PLAT-1010 | import lucid charts using quill

### PR DESCRIPTION
**Pull request by bognix**

### CODE REVIEW

File: resources/assets/js/quill/lucidChartFormat.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+		this.getPublicUrl(value).then(value => {
+			node.setAttribute('src', value);
+		});
+
+		return node;
+	}
+
+	static async getPublicUrl(value) {
+		const response = await axios.post(getApiUrl('lucid_charts'), {
```
Comment (bognix) [inline resources/assets/js/quill/lucidChartFormat.js]: error handling.

File: app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+use App\Http\Controllers\Api\ApiController;
+use App\Http\Requests\PostLucidChart;
+use Illuminate\Http\Request;
+use Facades\Lib\Bethink\Bethink;
+use Facades\Lib\LucidChart\LucidChart;
+
+class LucidChartsApiController extends ApiController {
+	public function __construct(Request $request) {
+		parent::__construct($request);
+		$this->resourceName = config('papi.resources.lucid-charts');
```
Comment (kvas-damian) [inline app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php]: I think it's not required while we don't use `transformAndRespond` or other `ApiController` methods

File: lib/LucidChart/LucidChart.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+	public $assetPath;
+	public $chartId;
+
+	public function createFromChartId($chartId) {
+		$this->chartId = $chartId;
+		$this->assetPath = "charts/{$this->chartId}.png";
+
+		$urlFormatted = sprintf(self::LUCID_URL_PATTERN, $this->chartId, self::LUCID_IMAGE_SIZE);
+		$image = Image::make($urlFormatted)->stream('png');
```
Comment (kvas-damian) [inline lib/LucidChart/LucidChart.php]: Did we check if they provide a better API for fetching the thumbnail now?

File: resources/assets/js/quill/lucidChartFormat.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+		return super.sanitize(response.data.public_path);
+	}
+}
+
+LucidChart.blotName = 'lucidChart';
+LucidChart.className = 'ql-lucidChart';
+LucidChart.tagName = 'IMG';
+
+const ToolbarIcon = '<svg viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg"><path d="M1792 1248v320q0 40-28 68t-68 28h-320q-40 0-68-28t-28-68v-320q0-40 28-68t68-28h96v-192h-512v192h96q40 0 68 28t28 68v320q0 40-28 68t-68 28h-320q-40 0-68-28t-28-68v-320q0-40 28-68t68-28h96v-192h-512v192h96q40 0 68 28t28 68v320q0 40-28 68t-68 28h-320q-40 0-68-28t-28-68v-320q0-40 28-68t68-28h96v-192q0-52 38-90t90-38h512v-192h-96q-40 0-68-28t-28-68v-320q0-40 28-68t68-28h320q40 0 68 28t28 68v320q0 40-28 68t-68 28h-96v192h512q52 0 90 38t38 90v192h96q40 0 68 28t28 68z"/></svg>';
```
Comment (kvas-damian) [inline resources/assets/js/quill/lucidChartFormat.js]: At some point, we should find a better way to include such svgs.

File: lib/LucidChart/LucidChart.php (hunk 3 above)
Comment (bognix) [reply lib/LucidChart/LucidChart.php]: Yes, they don't

File: resources/assets/js/quill/lucidChartFormat.js (hunk 4 above)
Comment (bognix) [reply resources/assets/js/quill/lucidChartFormat.js]: yes, in the future.

File: resources/assets/js/quill/lucidChartFormat.js (hunk 1 above)
Comment (bognix) [reply resources/assets/js/quill/lucidChartFormat.js]: not much error handling I can do here... Unfortunately this happens totally outside vue application and adding here our Vue alerts would require a bit different approach to creating the format.

Comment (kvas-damian) [reply resources/assets/js/quill/lucidChartFormat.js]: Maybe a simple alert of modal? Or maybe there is a basic error handling implemented in Quill?

File: app/Http/Requests/PostLucidChart.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+class PostLucidChart extends FormRequest
+{
+	/**
+	 * Determine if the user is authorized to make this request.
+	 *
+	 * @return bool
+	 */
+	public function authorize()
+	{
+		return true;
```
Comment (rogatty) [inline app/Http/Requests/PostLucidChart.php]: Admins only?

File: lib/LucidChart/LucidChart.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+class LucidChart
+{
+	const LUCID_URL_PATTERN = 'https://www.lucidchart.com/documents/thumb/%s/0/0/NULL/%d';
+	const LUCID_IMAGE_SIZE = 2000;
+
+	public $assetPath;
+	public $chartId;
+
+	public function createFromChartId($chartId) {
```
Comment (rogatty) [inline lib/LucidChart/LucidChart.php]: `static` is missing

File: app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+use App\Http\Controllers\Api\ApiController;
+use App\Http\Requests\PostLucidChart;
+use Facades\Lib\Bethink\Bethink;
+use Facades\Lib\LucidChart\LucidChart;
+
+class LucidChartsApiController extends ApiController {
+	public function post(PostLucidChart $request) {
+		$chartId = $request->chart_id;
+		$lucidChart = LucidChart::createFromChartId($chartId);
```
Comment (rogatty) [inline app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php]: It's not a static function.

File: resources/assets/js/quill/lucidChartFormat.js (hunk 1 above)
Comment (bognix) [reply resources/assets/js/quill/lucidChartFormat.js]: Done, I added our alerts.

File: app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php (hunk 7 above)
Comment (bognix) [reply app/Http/Controllers/Api/PrivateApi/LucidChartsApiController.php]: It's called using the Facade

**Pull request merged**